proposals/22866-unified-queue-implementation.md

```diff
@@ -0,0 +1,80 @@
+	* Setup experience references the `execution_id` of `host_software_installs`, should it bypass the unified queue as this _needs_ to happen during setup, not blocked on any other item that might be in the queue? https://github.com/fleetdm/fleet/blob/b4a5a1fb49666dd3b10cfd11ccf26190ad9d2902/server/datastore/mysql/setup_experience.go#L312
+
+- `host_vpp_software_installs` is used during the "pending" phase:
+	* Filter hosts by software install status : https://github.com/fleetdm/fleet/blob/b4a5a1fb49666dd3b10cfd11ccf26190ad9d2902/server/datastore/mysql/software_installers.go#L705
+	* List host software that is available for install : https://github.com/fleetdm/fleet/blob/b4a5a1fb49666dd3b10cfd11ccf26190ad9d2902/server/datastore/mysql/software.go#L2290
+	* Get summary of VPP software installs: https://github.com/fleetdm/fleet/blob/b4a5a1fb49666dd3b10cfd11ccf26190ad9d2902/server/datastore/mysql/vpp.go#L72
+
+### DB schema changes and migrations
+
+- Entries in the unified queue for a given host will need to be deleted if that host is deleted: https://github.com/fleetdm/fleet/blob/85dd21267700e76ee8fe184ddef2dc54692afe43/server/datastore/mysql/hosts.go#L563
+- Should secondary data be created when creating the unified queue pending record entry? E.g. the `script_contents` row for a script in a software install or script execution request?
+	* I'd say yes, otherwise we could end up taking lots of storage place in this table, defeating the purpose of the `script_contents` deduplication table.
+	* This means that if the script contents are updated, the pending row should be deleted (which I believe is how it works today).
+
+The new table will need to have at least those columns:
+- uuid (unique for each action, will allow cancelation support)
+- host_id (but without FK for performance reasons, as per our DB performance patterns)
+- actor id, email, name and avatar (copied into the table to persist even if user is deleted? a bit like we do for `activities`?)
+- fleet-initiated vs user-initiated? (e.g. a disk encryption MDM command or OS upgrade is fleet-initiated via Fleet settings, while deploying a custom profile is user-initiated by the user that added the profile to the team)
+- activity "type" (MDM command, script execution, software install, software uninstall, etc.)
+- a creation time and a priority column to allow for easy re-ordering (by default all are at the same prority so ordering could be `priority DESC, created_at ASC`)
+- an arbitrary JSON payload for the rest of the activity-specific payload, promoting columns as needed to support the "pending" behaviour that we have today via the other tables
+- a version number to allow for data migration of the JSON payload field (e.g. when changing that structure, a DB migration could use this version number to migrate it to the latest version, and then increment the version number)
```

Comment:
I'm not sure this is necessary - if we change the JSON payload for a Fleet release, then we need to migrate existing payloads to that new one. So whatever is in the table is always at the latest "version", I don't think we need to track that version number. That is, any payload in the table will never be in an older version than the current (now previous, since we're making changes to it) one.